xla/service/gpu/fusions/triton/triton_fusion_emitter.cc

```diff
@@ -1491,31 +1491,31 @@ class MatMulEmitterHelper {
 
   std::vector<const HloInstruction*> EpiloguePostOrderTransitiveOperands(
       const HloInstruction* root) {
-    // Collect all instructions of the dot's output scope.
-    absl::flat_hash_set<const HloInstruction*> to_order;
-    {
-      std::queue<const HloInstruction*> to_add;
-      if (root != dot_instr_) {
-        to_add.push(root);
-      }
-      while (!to_add.empty()) {
-        const HloInstruction* current = to_add.front();
-        for (const HloInstruction* operand : current->operands()) {
-          if (!to_order.contains(operand)) {
-            if (operand != dot_instr_) {
-              to_add.push(operand);
+    absl::flat_hash_set<const HloInstruction*> visited;
+    absl::flat_hash_set<const HloInstruction*> current_stack;
+
+    std::function<void(const HloInstruction*)> traverse_graph =
+        [&](const HloInstruction* instr) {
+          if (!instr || instr == dot_instr_ || visited.contains(instr)) {
+            return;
+          }
+          // Detects cycles or self-referential nodes.
```

Comment:
As discussed, there is no intention to detect cycles or self-references - that would be invalid HLO. The CHECK in this code was rather to ensure that the traversal is implemented correctly and it never adds the same node to to_order twice.